pyvelox/complex.cpp

```diff
@@ -0,0 +1,160 @@
+    // types which are supported then the Type tree cannot be constructed
+    if (v.kind() == TypeKind::UNKNOWN || v.kind() == TypeKind::INVALID) {
+      throw std::invalid_argument(
+          "Non-null variant has unknown or invalid kind");
+    }
+
+    switch (v.kind()) {
+      case TypeKind::ARRAY: {
+        counter.children.resize(1);
+        auto asArray = v.array();
+        TypePtr childType = createType(TypeKind::UNKNOWN, {});
+        for (const auto& element : asArray) {
+          constructType(childType, element, counter.children[0]);
+        }
+
+        // if child's type still remains Unknown, implies all the
+        // elements in the array are actually NULL
+        if (childType->kind() == TypeKind::UNKNOWN) {
+          throw py::value_error("Cannot construct array with all None values");
+        }
+        checkOrAssignType(type, [&childType]() {
+          return createType<TypeKind::ARRAY>({childType});
+        });
+        break;
+      }
+
+      default: {
+        checkOrAssignType(type, [&v]() { return createScalarType(v.kind()); });
+        break;
+      }
+    }
+  }
+}
+
+static void insertVariantIntoVector(
+    TypeKind& typeKind,
+    const variant& v,
+    VectorPtr& vector,
+    ElementCounter& counter) {
+  if (v.isNull()) {
+    vector->setNull(counter.insertedElements, true);
+  } else {
+    switch (typeKind) {
+      case TypeKind::ARRAY: {
+        auto asArray = vector->as<ArrayVector>();
+        asArray->elements()->resize(counter.children[0].totalElements);
+        const std::vector<variant>& elements = v.array();
+        vector_size_t offset = 0;
+        if (counter.insertedElements != 0) {
+          offset = asArray->offsetAt(counter.insertedElements - 1) +
+              asArray->sizeAt(counter.insertedElements - 1);
+        }
+        asArray->setOffsetAndSize(
+            counter.insertedElements, offset, elements.size());
+        for (const variant& elt : elements) {
+          auto elt_type = elt.kind();
+          insertVariantIntoVector(
+              elt_type, elt, asArray->elements(), counter.children[0]);
+        }
+
+        break;
+      }
+      default: {
+        VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+            setElementInFlatVector,
+            typeKind,
+            counter.insertedElements,
+            v,
+            vector);
+        break;
+      }
+    }
+  }
+  counter.insertedElements += 1;
+}
+
+VectorPtr variantsToVector(
+    const std::vector<variant>& variants,
+    velox::memory::MemoryPool* pool) {
+  ElementCounter counter;
+  TypePtr type = createType(TypeKind::UNKNOWN, {});
+  for (const auto& variant : variants) {
+    constructType(type, variant, counter);
+  }
+  VectorPtr resultVector =
+      BaseVector::create(std::move(type), variants.size(), pool);
+  for (const variant& v : variants) {
+    auto typeKind = v.kind();
+    insertVariantIntoVector(typeKind, v, resultVector, counter);
```

Comment:
if we the typekind from v, why even pass it to insertVariantIntoVector, can it not infer that from v ?

Reply:
correct! thanks for pointing that out, made the change.